data_model/src/events/data/events.rs

```diff
@@ -0,0 +1,269 @@
+                Self::Created(id)
+                | Self::Deleted(id)
+                | Self::AuthenticationAdded(id)
+                | Self::AuthenticationRemoved(id)
+                | Self::PermissionAdded(id)
+                | Self::PermissionRemoved(id)
+                | Self::MetadataInserted(id)
+                | Self::MetadataRemoved(id) => id,
+            }
+        }
+    }
+}
+
+mod domain {
+    //! This module contains `DomainEvent` and its impls
+
+    use super::*;
+
+    /// Domain Event
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    #[non_exhaustive]
+    #[allow(missing_docs)]
+    pub enum DomainEvent {
+        Account(AccountEvent),
+        AssetDefinition(AssetDefinitionEvent),
+        Created(DomainId),
+        Deleted(DomainId),
+        MetadataInserted(DomainId),
+        MetadataRemoved(DomainId),
+    }
+
+    impl IdTrait for DomainEvent {
+        type Id = DomainId;
+
+        fn id(&self) -> &DomainId {
+            match self {
+                Self::Account(account) => &account.id().domain_id,
+                Self::AssetDefinition(asset_definition) => &asset_definition.id().domain_id,
+                Self::Created(id)
+                | Self::Deleted(id)
+                | Self::MetadataInserted(id)
+                | Self::MetadataRemoved(id) => id,
+            }
+        }
+    }
+}
+
+mod trigger {
+    //! This module contains `TriggerEvent` and its impls
+
+    use super::*;
+
+    /// Trigger Event
+    #[derive(Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    #[non_exhaustive]
+    #[allow(missing_docs)]
+    pub enum TriggerEvent {
+        Created(TriggerId),
+        Deleted(TriggerId),
+        Extended(TriggerId),
+        Shortened(TriggerId),
+    }
+
+    impl IdTrait for TriggerEvent {
+        type Id = TriggerId;
+
+        fn id(&self) -> &TriggerId {
+            match self {
+                Self::Created(id)
+                | Self::Deleted(id)
+                | Self::Extended(id)
+                | Self::Shortened(id) => id,
+            }
+        }
+    }
+}
+
+/// World event
+///
+/// Does not participate in `Event`, but useful for events warranties when modifying `wsv`
+#[derive(
+    Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, FromVariant, IntoSchema,
+)]
+#[allow(missing_docs)]
+pub enum WorldEvent {
+    Domain(domain::DomainEvent),
+    Peer(peer::PeerEvent),
+
+    #[cfg(feature = "roles")]
+    Role(role::RoleEvent),
+}
+
+/// Event
+#[derive(
+    Clone, PartialEq, Eq, Debug, Decode, Encode, Deserialize, Serialize, FromVariant, IntoSchema,
+)]
+pub enum Event {
+    /// Domain event
+    Domain(domain::DomainEvent),
+    /// Peer event
+    Peer(peer::PeerEvent),
+    /// Role event
+    #[cfg(feature = "roles")]
+    Role(role::RoleEvent),
+    /// Account event
+    Account(account::AccountEvent),
```

Comment:
For example, as per previous comment, it seems to me that sometimes account events are going to be under `Event::Account` and sometimes under `Event::Domain::Account`. 

In fact, having looked more thoroughly, account events will always be produced under `Event::Domain::Account` entirely skipping the `Event::Account` variant. You could remove this variant(which I wouldn't recommend), or you could refactor as I suggested in the other comment. The same goes for asset events

Reply:
Why do you think, that `Event::Account` will be skipped?
When something was happened with account, then at least two events will happen: `DomainEvent::Account` and `Event::Account`.
Take a look at `modify_world()`:

```rust
// ...
match event {
    WorldEvent::Domain(domain_event) => {
        match &domain_event {
            DomainEvent::Account(account_event) => {
                // ...
                // Here we create `Event::AccountEvent`
                events.push(DataEvent::Account(account_event.clone()));
            }
            // ...
            // Here we create `Event::DomainEvent::Account`
            events.push(DataEvent::Domain(domain_event));
        }
}
self.produce_events(events);
// ...
```

Reply:
is there a case where these two events happen independently? Or the always come in tandem?